heron/ckptmgr/src/java/com/twitter/heron/ckptmgr/CheckpointManagerServer.java

```diff
@@ -0,0 +1,297 @@
+    } else if (request instanceof CheckpointManager.SaveInstanceStateRequest) {
+      handleSaveInstanceStateRequest(
+          rid, channel, (CheckpointManager.SaveInstanceStateRequest) request);
+    } else if (request instanceof CheckpointManager.GetInstanceStateRequest) {
+      handleGetInstanceStateRequest(
+          rid, channel, (CheckpointManager.GetInstanceStateRequest) request);
+    } else if (request instanceof CheckpointManager.CleanStatefulCheckpointRequest) {
+      handleCleanStatefulCheckpointRequest(
+          rid, channel, (CheckpointManager.CleanStatefulCheckpointRequest) request);
+    } else {
+      LOG.severe("Unknown kind of request: " + request.getClass().getName());
+    }
+  }
+
+  protected void handleCleanStatefulCheckpointRequest(
+      REQID rid,
+      SocketChannel channel,
+      CheckpointManager.CleanStatefulCheckpointRequest request
+  ) {
+    LOG.info("Got a clean request from " + request.toString() + " host:port "
+        + channel.socket().getRemoteSocketAddress());
+
+    boolean res = checkpointsBackend.dispose(request, topologyName);
+    Common.StatusCode statusCode = res ? Common.StatusCode.OK : Common.StatusCode.NOTOK;
+
+    CheckpointManager.CleanStatefulCheckpointResponse.Builder responseBuilder =
+        CheckpointManager.CleanStatefulCheckpointResponse.newBuilder();
+    responseBuilder.setStatus(Common.Status.newBuilder().setStatus(statusCode));
+
+    if (res) {
+      LOG.info("Dispose checkpoint successful");
+    } else {
+      LOG.info("Dispose checkpoint not successful");
+    }
+
+    sendResponse(rid, channel, responseBuilder.build());
+  }
+
+  protected void handleTMasterRegisterRequest(
+      REQID rid,
+      SocketChannel channel,
+      CheckpointManager.RegisterTMasterRequest request
+  ) {
+    LOG.info("Got a register request from TMaster host:port "
+        + channel.socket().getRemoteSocketAddress());
+
+    CheckpointManager.RegisterTMasterResponse.Builder responseBuilder =
+        CheckpointManager.RegisterTMasterResponse.newBuilder();
+
+    if (!request.getTopologyName().equals(topologyName)) {
+      LOG.severe("The register message was from a different topology: "
+          + request.getTopologyName());
+      responseBuilder.setStatus(Common.Status.newBuilder().setStatus(Common.StatusCode.NOTOK));
+    } else if (!request.getTopologyId().equals(topologyId)) {
+      LOG.severe("The register message was from a different topology id: "
+          + request.getTopologyName());
+      responseBuilder.setStatus(Common.Status.newBuilder().setStatus(Common.StatusCode.NOTOK));
+    } else if (connection != null) {
+      // TODO(mfu): Should we do this?
+      LOG.warning("We already have an active connection from the tmaster "
+          + "Closing existing connection...");
+
+      try {
+        connection.close();
+      } catch (IOException e) {
+        throw new RuntimeException("Failed to close connection from: "
```

Comment:
See other comment about not using `RuntimeExcepion` and not swallowing exceptions.